Revert Envoy xDS updates on regeneration failure

rlenglet opened the pull request:
Extend the xDS interfaces to return a revert function for every xDS cache update.
In `regenerateBPF`, maintain a stack of "revert" functions to revert the side-effects of the regeneration on failure. Also maintain a list of "finalize" functions that are executed only if regeneration completes successfully, to complete actions that are not reversible.
Revert xDS network policy changes and proxy redirect addition/removal on regeneration failure.

The effects of this PR is that all the side-effects on L7 proxies of an endpoint's BPF regeneration are reverted on regeneration failure, including all Envoy-related state, esp. the xDS cache updates.
No other regeneration side-effects are reverted. Notably, changes to the BPF policy map are not reverted, which can result in traffic loss if the policy map ends up not being consistent with the configuration of the L7 proxies after reverting.

Fixes: #5317
Signed-off-by: Romain Lenglet <romain@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5451)
<!-- Reviewable:end -->

rlenglet committed:
xds: Return a revert function on every cache update

Signed-off-by: Romain Lenglet <romain@covalent.io>

jrajahalme commented on pkg/envoy/xds/cache.go:
```diff
@@ -105,27 +109,49 @@ func (c *Cache) tx(typeURL string, upsertedResources map[string]proto.Message, d
+				if revertUpsertedResources == nil {
+					revertUpsertedResources = make(map[string]proto.Message, len(upsertedResources)+len(deletedNames))
+				}
+				revertUpsertedResources[name] = oldV.resource
```
I'd call this variable `revertChangedResources`...

rlenglet replied:
This map will be passed as parameter `upsertedResources`, so I wanted to match that name.

jrajahalme commented on pkg/envoy/xds/cache.go:
```diff
@@ -105,27 +109,49 @@ func (c *Cache) tx(typeURL string, upsertedResources map[string]proto.Message, d
+				if revertDeletedNames == nil {
+					revertDeletedNames = make([]string, 0, len(upsertedResources))
+				}
+				revertDeletedNames = append(revertDeletedNames, name)
```
And this one `deleteInsertedNames`?

rlenglet replied:
Likewise, this slice will be passed as the `deletedNames` parameter, and I wanted to match that name.

jrajahalme reviewed (approved):
LGTM

rlenglet committed:
endpoint: Implement stack of endpoint state revert functions

Maintain a stack of revert functions in regenerateBPF to revert the
side-effects of the regeneration on failure.

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
endpoint: Revert xDS network policy changes on regeneration failure

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
endpoint: Revert listener addition/removal on configuration failure

Signed-off-by: Romain Lenglet <romain@covalent.io>

aanm commented on pkg/envoy/xds/ack_test.go:
```diff
@@ -229,3 +229,98 @@ func (s *AckSuite) TestDeleteMultipleNodes(c *C) {
+	defer comp.Complete()
+
+	// Insert.
+	revert := acker.Upsert(typeURL, resources[0].Name, resources[0], []string{node0}, comp)
```
Can you add a test that runs `upsert` 2 times and you want to revert the first upsert?

rlenglet replied:
Done.

aanm commented on pkg/envoy/xds/cache.go:
```diff
@@ -105,27 +109,49 @@ func (c *Cache) tx(typeURL string, upsertedResources map[string]proto.Message, d
 			v.resource = value
 			c.resources[k] = v
 		}
+		if force {
```
You can put this outside of this for loop and outside of the next for loop and have a if statement that does
```
if len(upsertedResources) != 0 || len(deletedNames) != 0{
cacheIsUpdated = true
}
```

rlenglet replied:
That's not the correct logic. See that if `!force`, `cacheIsUpdated` is set to `true` only when:
- A resource to update is inserted or different from the current version of the resource in the cache.
- A resource to delete actually already exists in the cache.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -450,6 +450,18 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
+	defer func() {
+		if reterr != nil {
+			e.Logger().Error("Restoring endpoint state after BPF regeneration failed")
+			if err := e.Revert(&revertStack); err != nil {
```
Maybe these changes is in the wrong commit? I don't see a `e.Push`

ianvernon replied:
@aanm I think this is just the framework part of it - with just this commit, there aren't any functional changes since nothing is pushed yet.

rlenglet replied:
Correct, this is built in multiple commits.

ianvernon commented on pkg/endpoint/revert.go:
```diff
@@ -0,0 +1,54 @@
+
+// Push pushes the given RevertFunc on top of this stack. If the function is
+// nil, it is ignored.
+func (s *RevertStack) Push(revertFunc RevertFunc) {
```
unit tests ?

ianvernon commented on pkg/endpoint/revert.go:
```diff
@@ -0,0 +1,54 @@
+// Copyright 2018 Authors of Cilium
```
There's nothing wrong with this being in the endpoint package since endpoint is only consumer, but I could see this type being in its own separate package  (apart from the endpoint logger being used). It's pretty small though, so that's an argument in keeping it where it is now.

rlenglet replied:
Done.

ianvernon commented on pkg/endpoint/revert.go:
```diff
@@ -52,3 +52,41 @@ func (e *Endpoint) Revert(s *RevertStack) error {
+
+// Append appends the given FinalizeFunc at the end of this list. If the
+// function is nil, it is ignored.
+func (s *FinalizeList) Append(finalizeFunc FinalizeFunc) {
```
Unit tests?

ianvernon commented on pkg/proxy/proxy.go:
```diff
@@ -146,12 +147,25 @@ func (p *Proxy) allocatePort() (uint16, error) {
+// RevertFunc is a function returned by a successful function call, which
+// reverts the side-effects of the initial function call. A call that returns
+// an error should return a nil RevertFunc.
+type RevertFunc func() error
```
if RevertFunc and FinalizeFunc were in their own packages, you wouldn't have to duplicate the types in proxy / endpoint packages.

rlenglet replied:
Good idea, it's much cleaner. Done.

ianvernon commented on pkg/proxy/proxy.go:
```diff
@@ -172,45 +186,61 @@ func (p *Proxy) CreateOrUpdateRedirect(l4 *policy.L4Filter, id string, localEndp
+	revertFuncs := make([]RevertFunc, 0, 2)
+	revertFunc = func() error {
+		for i := len(revertFuncs) - 1; i >= 0; i-- {
+			if err := revertFuncs[i](); err != nil {
```
Shouldn't we try to run all revert funcs if possible? And then return a list of errors? I guess we will have inconsistent state either way, though :/

rlenglet replied:
There are typically dependencies between the revert functions, so that's not possible.

ianvernon reviewed (commented):
LGTM overall, just a few stylistic nitpicks. It would be nice to document what exactly the behavior of the datapath is once we hit these cases. I.e., if the reverts are successful, can we assume that the endpoint will have full connectivity as if nothing happened? I ask because of how proxy-ports are integrated into the PolicyMap. The main fix here is to not have Envoy itself be in a bad state, as opposed to having the datapath be OK for endpoints. Is that correct?

jrajahalme commented on pkg/endpoint/bpf.go:
```diff
@@ -263,16 +270,26 @@ func (e *Endpoint) addNewRedirectsFromMap(owner Owner, m policy.L4PolicyMap, des
 				}
 				e.realizedRedirects[proxyID] = redirectPort
+				revertStack.Push(func() error {
+					delete(e.realizedRedirects, proxyID)
```
Does it matter that `proxyID` could have been in the map previously and now it is just deleted?

rlenglet replied:
Good catch. I updated to only delete the `proxyID` from `realizedRedirects` if it wasn't there before regeneration.

jrajahalme reviewed (commented):
Will review the xds pieces after rebase.

rlenglet committed:
endpoint: Move the revert types and logic into pkg/revert

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
endpoint: In revert, only delete from realizedRedirects if necessary

Signed-off-by: Romain Lenglet <romain@covalent.io>

jrajahalme reviewed (approved):
Went through the rebased code, still looking good.

rlenglet committed:
endpoint: Add proxy revert logging

Signed-off-by: Romain Lenglet <romain@covalent.io>
